### PR TITLE
[FLINK-6018] Add tests for KryoSerializer restore with registered types

### PR DESCRIPTION
**Pull request by aljoscha**
This is the result of the discussion in #3534. 

I changed `TypeSerializer.isCompatibleWith()` to `TypeSerializer.canRestoreFrom` because the relation is not necessarily symmetric.
 
I added a `KryoSerializer.canRestoreFrom()` that only allows restoring when we previously didn't have registered types/serializers.

I added a whole bunch of tests in `StateBackendTestBase`, this should be review most thoroughly.

R: @StephanEwen and @tzulitai because this probably is interesting with the serialiser update story that you're working on.

### CODE REVIEW

**Comment by tzulitai** [comment]
+1 to the rename. I was thinking the same here :-)

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 1]
```diff
@@ -82,13 +90,19 @@
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
+import com.esotericsoftware.kryo.serializers.JavaSerializer;
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Can we use `o.a.f.api.java.typeutils.kryo.JavaSerializer` instead (see https://issues.apache.org/jira/browse/FLINK-6025)? 
I don't think it really matters for the tests, but we probably should start using it instead of Kryo's because of the bug in case anything unexpected pops up.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 2]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+		assertTrue(pojoType.createSerializer(env.getExecutionConfig()) instanceof KryoSerializer);
+
+		pojoType.createSerializer(env.getExecutionConfig());
+
+		ValueStateDescriptor<TestPojo> kvId = new ValueStateDescriptor<>("id", pojoType);
+
+		ValueState<TestPojo> state = backend.getPartitionedState(VoidNamespace.INSTANCE, VoidNamespaceSerializer.INSTANCE, kvId);
+
+		// backends that eagerly serializer will fail when updating state, others will
+		// fail only when performing the snapshot
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: nitpick:
It'll be helpful if the comment mentioned that the exception is thrown because the `ExceptionThrowingTestSerializer` is expected to be used; got a tiny bit confused here ;-)

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 3]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+			runSnapshot(backend.snapshot(682375462378L, 2, streamFactory, CheckpointOptions.forFullCheckpoint()));
+		} catch (ExpectedKryoTestException e) {
+			numExceptions++;
+		}
+
+		assertEquals("Didn't see the expected Kryo exception.", 1, numExceptions);
+	}
+
+	@Test
+	public void testBackendUsesRegisteredKryoDefaultSerializerUsingGetOrCreate() throws Exception {
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Same comments as in the `testBackendUsesRegisteredKryoDefaultSerializer` test

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 4]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+			runSnapshot(backend.snapshot(682375462378L, 2, streamFactory, CheckpointOptions.forFullCheckpoint()));
+		} catch (ExpectedKryoTestException e) {
+			numExceptions++;
+		}
+
+		assertEquals("Didn't see the expected Kryo exception.", 1, numExceptions);
+	}
+
+	@Test
+	public void testBackendUsesRegisteredKryoSerializer() throws Exception {
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Same comments as in the `testBackendUsesRegisteredKryoDefaultSerializer` test.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 5]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+
+		// make sure that we are in fact using the KryoSerializer
+		assertTrue(pojoType.createSerializer(env.getExecutionConfig()) instanceof KryoSerializer);
+
+		pojoType.createSerializer(env.getExecutionConfig());
+
+		ValueStateDescriptor<TestPojo> kvId = new ValueStateDescriptor<>("id", pojoType);
+
+		ValueState<TestPojo> state = backend.getOrCreateKeyedState(VoidNamespaceSerializer.INSTANCE, kvId);
+		assert state instanceof InternalValueState;
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Why not use `assertTrue` here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 6]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: `assertTrue`

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 7]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+			runSnapshot(backend.snapshot(682375462378L, 2, streamFactory, CheckpointOptions.forFullCheckpoint()));
+		} catch (ExpectedKryoTestException e) {
+			numExceptions++;
+		}
+
+		assertEquals("Didn't see the expected Kryo exception.", 1, numExceptions);
+	}
+
+	@Test
+	public void testBackendUsesRegisteredKryoSerializerUsingGetOrCreate() throws Exception {
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Same comments as in the testBackendUsesRegisteredKryoDefaultSerializer test.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 8]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+		env.getExecutionConfig()
+				.addDefaultKryoSerializer(TestPojo.class, (Class) CustomKryoTestSerializer.class);
+
+		backend = restoreKeyedBackend(IntSerializer.INSTANCE, snapshot, env);
+
+		snapshot.discardState();
+
+		// re-initialize to ensure that we create the Kryo Serializer from scratch, otherwise
+		// initializeSerializerUnlessSet would not pick up our new config
+		kvId = new ValueStateDescriptor<>("id", pojoType);
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Why is this necessary here, but not in `testKryoRegisteringRestoreResilienceWithRegisteredType`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 9]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+
+		KeyGroupsStateHandle snapshot = runSnapshot(backend.snapshot(
+				682375462378L,
+				2,
+				streamFactory,
+				CheckpointOptions.forFullCheckpoint()));
+
+		backend.dispose();
+
+		env.getExecutionConfig().registerKryoType(TestPojo.class);
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Should we also reinitialize `kvId` after this, just to be sure?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 10]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+				682375462378L,
+				2,
+				streamFactory,
+				CheckpointOptions.forFullCheckpoint()));
+
+		backend.dispose();
+
+		// cast because our test serializer is not typed to TestPojo
+		env.getExecutionConfig()
+				.addDefaultKryoSerializer(TestPojo.class, (Class) CustomKryoTestSerializer.class);
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Reinitialize `kvId` after this, also?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 11]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+
+		state = backend.getPartitionedState(VoidNamespace.INSTANCE, VoidNamespaceSerializer.INSTANCE, kvId);
+
+		backend.setCurrentKey(1);
+		state.value();
+	}
+
+
+	/**
+	 * Verify that we can restore a snapshot that was done with without registered serializers
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: "with without" redundancy --> "without", I think ;-)?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 12]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+		assertEquals(state.value(), new TestPojo("u1", 1));
+
+		backend.setCurrentKey(2);
+		assertEquals(state.value(), new TestPojo("u2", 2));
+
+		backend.dispose();
+	}
+
+	/**
+	 * Verify that we can restore a snapshot that was done with without registered default
```
Comment (tzulitai) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: "with without" redundancy --> "without", I think ;-)?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 13]
```diff
@@ -162,22 +176,422 @@ protected CheckpointStreamFactory createStreamFactory() throws Exception {
...
+		// cast because our test serializer is not typed to TestPojo
+		env.getExecutionConfig()
+				.addDefaultKryoSerializer(TestPojo.class, (Class) ExceptionThrowingTestSerializer.class);
+
+		TypeInformation<TestPojo> pojoType = new GenericTypeInfo<>(TestPojo.class);
+
+		// make sure that we are in fact using the KryoSerializer
+		assertTrue(pojoType.createSerializer(env.getExecutionConfig()) instanceof KryoSerializer);
+
+		pojoType.createSerializer(env.getExecutionConfig());
```
Comment (aljoscha) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Must have missed that when I cleaned up the code after experimenting ... 😓 Fixing.

**Comment by tzulitai** [comment]
@aljoscha I've been trying out this PR to also test some Kryo behaviours. I can address the above comments along the way and open a PR against this one :-)

**Comment by tzulitai** [comment]
Here's my follow-up fixes for this: https://github.com/tzulitai/flink/tree/FLINK-6018

It also includes a rebase for the merge conflict + test fixes.
We can continue with this after Aljoscha comes back from vacation :)

**Comment by aljoscha** [comment]
@tzulitai I would suggest you open a PR with two commits: mine and your changes on top. What do you think? Then I would close this PR.

**Comment by tzulitai** [comment]
@aljoscha sure, I'll do that.

**Comment by tzulitai** [comment]
Done. Re-opened at #3603.

**Comment by aljoscha** [comment]
Closing in favour of #3603
